provider/pkg/provider/handlers.go

```diff
@@ -0,0 +1,98 @@
+// Copyright 2021, Pulumi Corporation.
+
+package provider
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/imdario/mergo"
+	"github.com/pkg/errors"
+	"github.com/pulumi/pulumi-google-native/provider/pkg/resources"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/logging"
+	rpc "github.com/pulumi/pulumi/sdk/v3/proto/go"
+)
+
+type updateHandler interface {
+	Update(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error)
+}
+
+type updateHandlerFunc func(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error)
+
+func (u updateHandlerFunc) Update(inputs resource.PropertyMap, oldState resource.PropertyMap) (map[string]interface{}, error) {
+	return u(inputs, oldState)
+}
+
+func (p *googleCloudProvider) updateHandler(resourceTok string, req *rpc.UpdateRequest) updateHandler {
+	logging.V(9).Infof("UpdateHandler for %q", resourceTok)
+	switch resourceTok {
+	case "google-native:container/v1:Cluster", "google-native:container/v1beta1:Cluster":
+		return updateHandlerFunc(func(inputs, oldState resource.PropertyMap) (map[string]interface{}, error) {
+			res, ok := p.resourceMap.Resources[resourceTok]
+			if !ok {
+				return nil, errors.Errorf("resource %q not found", resourceTok)
+			}
+
+			updateBody := p.prepareAPIInputs(inputs, oldState, res.UpdateProperties)
+
+			// TODO: Cluster updates also include flattened members of the `networkConfig` map
+			// which is hard to represent with the current metadata format in place (sdkName
+			// can only be used to refer to fields at the top level).
+
+			var splitBody []map[string]interface{}
+			update, hasUpdate := updateBody["update"]
+			updateMap, isMap := update.(map[string]interface{})
+			if hasUpdate && isMap {
+				for k, v := range updateMap {
+					if strings.HasPrefix(k, "desired") {
+						splitBody = append(splitBody, map[string]interface{}{k: v})
```

Comment:
I don't quite get how `splitBody` works... Is it an array of maps, each map having exactly one key? Are we sending exactly one key in the update body? Will it keep other fields intact?

This seems very different from a "normal" update where we'd send the full payload of the updated resource, so it's worth documenting the unique behavior here.